config/src/main/java/org/springframework/security/config/annotation/web/configuration/OAuth2ClientConfiguration.java

```diff
@@ -91,23 +115,33 @@ public void addArgumentResolvers(List<HandlerMethodArgumentResolver> argumentRes
 		}
 
 		@Autowired(required = false)
-		public void setClientRegistrationRepository(List<ClientRegistrationRepository> clientRegistrationRepositories) {
+		void setClientRegistrationRepository(List<ClientRegistrationRepository> clientRegistrationRepositories) {
 			if (clientRegistrationRepositories.size() == 1) {
 				this.clientRegistrationRepository = clientRegistrationRepositories.get(0);
 			}
 		}
 
 		@Autowired(required = false)
-		public void setAuthorizedClientRepository(List<OAuth2AuthorizedClientRepository> authorizedClientRepositories) {
+		void setAuthorizedClientRepository(List<OAuth2AuthorizedClientRepository> authorizedClientRepositories) {
 			if (authorizedClientRepositories.size() == 1) {
 				this.authorizedClientRepository = authorizedClientRepositories.get(0);
 			}
 		}
 
 		@Autowired
-		public void setAccessTokenResponseClient(
+		void setAccessTokenResponseClient(
 				Optional<OAuth2AccessTokenResponseClient<OAuth2ClientCredentialsGrantRequest>> accessTokenResponseClient) {
 			accessTokenResponseClient.ifPresent(client -> this.accessTokenResponseClient = client);
 		}
 	}
+
+	@Configuration(proxyBeanMethods = false)
+	static class OAuth2ClientRegistrationsConfiguration {
+
+		@Autowired
+		@Qualifier(OAuth2ClientBeanNames.REST_OPERATIONS)
+		void configure(RestOperations restOperations) {
+			ClientRegistrations.setRestOperations(restOperations);
```

Comment:
Setting a static method from a bean definition is not a typical pattern I find in Spring. Instead, I'd expect to be able to provide the RestOperations to the builder. A few concerns around this design:

First, it is not easy to provide different ``RestOperation``s for different issuers in a thread safe way. The builders should ideally be thread safe. This is especially true, given the performance improvements being made to Spring and talks around making bean creation multi-threaded.

Another and more immediate concern is that it appears that there is likely a bean ordering issue here. What guarantees this will happen before `ClientRegistrations` is used to create a client registration?

Reply:
> Setting a static method from a bean definition is not a typical pattern I find in Spring.

Understood. This is the simplest solution I came up. We're kind of in a bind with this class since it's not designed to be a `@Bean` and instead a utility class.

> I'd expect to be able to provide the RestOperations to the builder

There is no builder in `ClientRegistrations`? I'm don't understand. Can you clarify what you mean here?

> Another and more immediate concern is that it appears that there is likely a bean ordering issue here

See test `OAuth2ClientConfigurationTests.loadContextWhenRestOperationsRegisteredThenClientRegistrationsUses()` as it passes.

Honestly, I'm not a fan of this static setter either and would be fine not exposing it. But then how do we address #7027? Do we introduce #5543 and deprecate `ClientRegistrations`?

Reply:
> There is no builder in ClientRegistrations? I don't understand. Can you clarify what you mean here?

Sorry for being unclear. An option would be to introduce a builder to ClientRegistrations that allows passing in the `RestOperations` that would be used.

> Honestly, I'm not a fan of this static setter either and would be fine not exposing it. But then how do we address #7027? Do we introduce #5543 and deprecate ClientRegistrations?

I think an option would be that Spring Security leverages the configured RestOperations to pass into the newly added builder in ClientRegistrations.

Reply:
> An option would be to introduce a builder to ClientRegistrations

I'm not sure how this would work @rwinch? Maybe I'm missing something.

`ClientRegistrations` was designed as a utility `final` class with `static` methods.

What would `ClientRegistrations.Builder.build()` return? An instance of `ClientRegistrations`?